File: docs/core/examples/ruby/batch_operations.rb
require 'chain'

chain = Chain::Client.new

key = chain.mock_hsm.keys.create
signer.add_key(key, chain.mock_hsm.signer_conn)

# snippet asset-builders
List<Asset.Builder> assetBuilders = Arrays.asList(
  chain.assets.create(
    alias: 'gold',
    root_xpubs: [key.xpub],
    quorum: 1,,
  chain.assets.create(
    alias: 'silver',
    root_xpubs: [key.xpub],
    quorum: 1,,
  chain.assets.create(
    alias: 'bronze',
    root_xpubs: [key.xpub],
    quorum: [0],
)
# endsnippet

# snippet asset-create-batch
BatchResponse<Asset> assetBatch = Asset.createBatch(client, assetBuilders)
# endsnippet

# snippet asset-create-handle-errors
for (int i = 0 i < assetBatch.size() i++) {
  if (assetBatch.isError(i)) {
    APIException error = assetBatch.errorsByIndex().get(i)
    puts('asset ' + i + ' error: ' + error)
  } else {
    Asset asset = assetBatch.successesByIndex().get(i)
    puts('asset ' + i + ' created, ID: ' + asset.id)
  }
}
# endsnippet

# snippet nondeterministic-errors
assetBuilders = Arrays.asList(
  chain.assets.create(
    alias: 'platinum',
    root_xpubs: [key.xpub],
    quorum: 1,,
  chain.assets.create(
    alias: 'platinum',
    root_xpubs: [key.xpub],
    quorum: 1,,
  chain.assets.create(
    alias: 'platinum',
    root_xpubs: [key.xpub],
    quorum: 1,
)
# endsnippet

assetBatch = Asset.createBatch(client, assetBuilders)

for (int i = 0 i < assetBatch.size() i++) {
  if (assetBatch.isError(i)) {
    APIException error = assetBatch.errorsByIndex().get(i)
    puts('asset ' + i + ' error: ' + error)
  } else {
    Asset asset = assetBatch.successesByIndex().get(i)
    puts('asset ' + i + ' created, ID: ' + asset.id)
  }
}

chain.accounts.create(
  alias: 'alice',
  root_xpubs: [key.xpub],
  quorum: 1,
)

chain.accounts.create(
  alias: 'bob',
  root_xpubs: [key.xpub],
  quorum: 1,
)

# snippet batch-build-builders
List<Transaction.Builder> txBuilders = Arrays.asList(
  chain.transactions.build do |b|
    b.issue
      asset_alias: 'gold',
      amount: 100,
    b.control_with_account
      account_alias: 'alice',
      asset_alias: 'gold',
      amount: 100,
    ),
  chain.transactions.build do |b|
    b.issue
      asset_alias: 'not-a-real-asset',
      amount: 100,
    b.control_with_account
      account_alias: 'alice',
      asset_alias: 'not-a-real-asset',
      amount: 100,
    ),
  chain.transactions.build do |b|
    b.issue
      asset_alias: 'silver',
      amount: 100,
    b.control_with_account
      account_alias: 'alice',
      asset_alias: 'silver',
      amount: 100,
    )
)
# endsnippet

# snippet batch-build-handle-errors
BatchResponse<Transaction.Template> buildTxBatch = Transaction.buildBatch(client, txBuilders)

for(Map.Entry<Integer, APIException> err : buildTxBatch.errorsByIndex().entrySet()) {
  puts('Error building transaction ' + err.getKey() + ': ' + err.getValue())
}
# endsnippet

# snippet batch-sign
BatchResponse<Transaction.Template> signTxBatch = signer.signBatch(buildTxBatch.successes())

for(Map.Entry<Integer, APIException> err : signTxBatch.errorsByIndex().entrySet()) {
  puts('Error signing transaction ' + err.getKey() + ': ' + err.getValue())
}
# endsnippet

# snippet batch-submit
BatchResponse<Transaction.SubmitResponse> submitTxBatch = Transaction.submitBatch(client, signTxBatch.successes())

for(Map.Entry<Integer, APIException> err : submitTxBatch.errorsByIndex().entrySet()) {
  puts('Error submitting transaction ' + err.getKey() + ': ' + err.getValue())
}

for(Map.Entry<Integer, Transaction.SubmitResponse> success : submitTxBatch.successesByIndex().entrySet()) {
  puts('' + success.getKey() + ' submitted, ID: ' + success.getValue().id)
}
# endsnippet

File: docs/core/examples/node/batchOperations.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let key

client.mockHsm.keys.create().then(_key => {
  key = _key
  signer.addKey(key.xpub, client.mockHsm.signerConnection)
}).then(() => {
  // snippet asset-builders
  const assetsToBuild = [{
    alias: 'gold',
    rootXpubs: [key.xpub],
    quorum: 1,
  }, {
    alias: 'silver',
    rootXpubs: [key.xpub],
    quorum: 1,
  }, {
    alias: 'bronze',
    rootXpubs: [key.xpub],
    quorum: 0,
  }]
  // endsnippet

  // snippet asset-create-batch
  const assetBatchPromise = client.assets.createBatch(assetsToBuild)
  // endsnippet

  return assetBatchPromise
}).then(assetBatch => {
  // snippet asset-create-handle-errors
  assetBatch.errors.forEach((err, index) => {
    console.log(`asset ${index} error: `)
    console.log(err)
  })

  assetBatch.successes.forEach((asset, index) => {
    console.log(`asset ${index} created, ID: ${asset.id}`)
  })
  // endsnippet
}).then(() => {
  // snippet nondeterministic-errors
  const assetsToBuild = [{
    alias: 'platinum',
    rootXpubs: [key.xpub],
    quorum: 1,
  }, {
    alias: 'platinum',
    rootXpubs: [key.xpub],
    quorum: 1,
  }, {
    alias: 'platinum',
    rootXpubs: [key.xpub],
    quorum: 1,
  }]
  // endsnippet

  return client.assets.createBatch(assetsToBuild)
}).then(assetBatch => {
  assetBatch.errors.forEach((err, index) => {
    console.log(`asset ${index} error: `)
    console.log(err)
  })

  assetBatch.successes.forEach((asset, index) => {
    console.log(`asset ${index} created, ID: ${asset.id}`)
  })
}).then(() => client.accounts.createBatch([
  { alias: 'alice', rootXpubs: [key.xpub], quorum: 1 },
  { alias: 'bob', rootXpubs: [key.xpub], quorum: 1 },
])).then(() => {
  // snippet batch-build-builders
  const transactionsToBuild = [
    (builder) => {
      builder.issue({assetAlias: 'gold', amount: 100})
      builder.controlWithAccount({accountAlias: 'alice', assetAlias: 'gold', amount: 100})
    },
    (builder) => {
      builder.issue({assetAlias: 'not-a-real-asset', amount: 100})
      builder.controlWithAccount({accountAlias: 'alice', assetAlias: 'not-a-real-asset', amount: 100})
    },
    (builder) => {
      builder.issue({assetAlias: 'silver', amount: 100})
      builder.controlWithAccount({accountAlias: 'alice', assetAlias: 'silver', amount: 100})
    }
  ]
  // endsnippet

  // snippet batch-build-handle-errors
  const buildBatchPromise = client.transactions.buildBatch(transactionsToBuild)
    .then((buildBatch) => {
      buildBatch.errors.forEach((err, index) => {
        console.log(`Error building transaction ${index} error: `)
        console.log(err)
      })

      return buildBatch
    })
  // endsnippet

  return buildBatchPromise
}).then((buildBatch) => {
  // snippet batch-sign
  const transactionsToSign = buildBatch.successes
  const signBatchPromise = signer.signBatch(transactionsToSign)
    .then((signBatch) => {
      signBatch.errors.forEach((err, index) => {
        console.log(`Error signing transaction ${index} error: `)
        console.log(err)
      })

      return signBatch
    })
  // endsnippet

  return signBatchPromise
}).then((signBatch) => {
  // snippet batch-submit
  const transactionsToSubmit = signBatch.successes
  const submitBatchPromise = client.transactions.submitBatch(transactionsToSubmit)
    .then((submitBatch) => {
      submitBatch.errors.forEach((err, index) => {
        console.log(`Error submitting transaction ${index} error: `)
        console.log(err)
      })

      submitBatch.successes.forEach((tx, index) => {
        console.log(`Transaction ${index} submitted, ID: ${tx.id}`)
      })

      return submitBatch
    })
  // endsnippet
}).catch(err =>
  process.nextTick(() => { throw err })
)
